x/stake/fee_distribution.go

```diff
@@ -5,6 +5,65 @@ import (
 )
 
 // Handle fee distribution to the validators and delegators
-func FeeHandler(ctx sdk.Context, tx sdk.Tx, fee sdk.Coins) {
+func (k Keeper) FeeHandler(ctx sdk.Context, collectedFees sdk.Coins) {
+	pool := k.GetPool(ctx)
+	params := k.GetParams(ctx)
 
+	// XXX determine
+	candidate := NewCandidate(addrs[0], pks[0], Description{})
+
+	// calculate the proposer reward
+	precommitPower := k.GetTotalPrecommitVotingPower(ctx)
+	toProposer := coinsMulRat(collectedFees, (sdk.NewRat(1, 100).Add(sdk.NewRat(4, 100).Mul(precommitPower).Quo(pool.BondedShares))))
+	candidate.ProposerRewardPool = candidate.ProposerRewardPool.Plus(toProposer)
+
+	toReservePool := coinsMulRat(collectedFees, params.ReservePoolFee)
+	pool.FeeReservePool = pool.FeeReservePool.Plus(toReservePool)
+
+	distributedReward := (collectedFees.Minus(toProposer)).Minus(toReservePool)
+	pool.FeePool = pool.FeePool.Plus(distributedReward)
+	pool.FeeSumReceived = pool.FeeSumReceived.Plus(distributedReward)
+	pool.FeeRecent = distributedReward
+
+	// lastly update the FeeRecent term
+	pool.FeeRecent = collectedFees
+
+	k.setPool(ctx, pool)
+}
+
+func coinsMulRat(coins sdk.Coins, rat sdk.Rat) sdk.Coins {
+	var res sdk.Coins
+	for _, coin := range coins {
+		coinMulAmt := rat.Mul(sdk.NewRat(coin.Amount)).Evaluate()
+		coinMul := sdk.Coins{{coin.Denom, coinMulAmt}}
+		res = res.Plus(coinMul)
```

Comment:
This is an inefficient way to implement this function (`O(n^2)` instead of `O(n)`, with `n` the number of fee coin denominations). We can just append `sdk.Coin` to the coins array.

Reply:
oh you mean we don't need to use `res = res.Plus(coinMul)`. Yeah good point